### DISCUSSION TITLE
brew can't set HOMEBREW_CURL_PATH to a newer version

### DISCUSSION THREAD

**Discussion by starrzy** [Homebrew/discussions#3558] [Getting started]
I can't use sudo, and I install brew in my home directory:` ~/.linuxbrew/bin/`.

After I export` ~/.linuxbrew/bin/` into PATH:

`export PATH=$PATH:~/.linuxbrew/bin/`

I executed `brew` and I met this error:
<img width="873" alt="Screen Shot 2022-08-23 at 4 32 47 PM" src="https://user-images.githubusercontent.com/37285589/186111753-4663e102-4948-4251-bb23-47c308ea3791.png">

Then I searched some solutions online, one of them told that `HOMEBREW_DEVELOPER=1
HOMEBREW_GIT_PATH=/path/to/your/git
HOMEBREW_CURL_PATH=/path/to/your/curl` is OK.

I tried, but the error still exists.
<img width="545" alt="Screen Shot 2022-08-23 at 4 38 28 PM" src="https://user-images.githubusercontent.com/37285589/186112994-256d4a46-1ce8-4e3c-9a8e-dea0d18731ec.png">
<img width="789" alt="Screen Shot 2022-08-23 at 4 39 07 PM" src="https://user-images.githubusercontent.com/37285589/186113122-8a0554a2-f8cd-4799-b9ed-5cedaab9204f.png">
<img width="838" alt="Screen Shot 2022-08-23 at 4 39 33 PM" src="https://user-images.githubusercontent.com/37285589/186113203-cafd8eee-ad91-4598-83cf-9e0a493d6848.png">

I don't know how to sovle it. Can anyone help me? Thanks!

**Comment by marionrobert** [comment]
Hello, I have the same error. Did you finally solve it please? I'm totally blocked...

**Comment by gromgit** [comment]
> Then I searched some solutions online, one of them told that `HOMEBREW_DEVELOPER=1 HOMEBREW_GIT_PATH=/path/to/your/git HOMEBREW_CURL_PATH=/path/to/your/curl` is OK.

That is more-or-less correct advice, you just used the wrong values. You should set both paths to the _full pathnames_ of each executable, not just the directory they're in, so:
```bash
HOMEBREW_CURL_PATH=/home/rzy/software/curl/bin/curl
HOMEBREW_GIT_PATH=/home/rzy/software/git/bin/git
```
Also, please paste text output _as text_, not screenshots. It's impossible to copy-and-paste graphical characters for testing purposes, for starters.

**Reply by marionrobert** [reply]
Hello @gromgit , thank you for your advices. However, it's still not working. 

**I set the variables in the .zshrc file:**
```
export HOMEBREW_CURL_PATH=/bin/curl
export HOMEBREW_GIT_PATH=/bin/git
```

Well, I am not sure about the path...

![set variables](https://user-images.githubusercontent.com/107509668/210813055-5003a76c-fec4-425b-b412-909da33eb83d.png)



Then, I ran again the command (/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)")
and I have still the same error.

```
 ~ /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
==> Found Git: ./bin/git
==> Found cURL: ./bin/curl
==> Setting HOMEBREW_DEVELOPER to use Git/cURL not in /usr/bin
==> Checking for `sudo` access (which may request your password)...
==> This script will install:
/home/linuxbrew/.linuxbrew/bin/brew
/home/linuxbrew/.linuxbrew/share/doc/homebrew
/home/linuxbrew/.linuxbrew/share/man/man1/brew.1
/home/linuxbrew/.linuxbrew/share/zsh/site-functions/_brew
/home/linuxbrew/.linuxbrew/etc/bash_completion.d/brew
/home/linuxbrew/.linuxbrew/Homebrew

Press RETURN/ENTER to continue or any other key to abort:
[sudo] password for marionrbt:
==> /usr/bin/sudo /bin/chown -R marionrbt:marionrbt /home/linuxbrew/.linuxbrew/Homebrew
==> Downloading and installing Homebrew...
remote: Enumerating objects: 27, done.
remote: Counting objects: 100% (26/26), done.
remote: Compressing objects: 100% (17/17), done.
remote: Total 27 (delta 17), reused 12 (delta 9), pack-reused 1
Unpacking objects: 100% (27/27), 18.01 KiB | 1.29 MiB/s, done.
From https://github.com/Homebrew/brew
   9e9b71bea..ea0b78638  master     -> origin/master
HEAD is now at ea0b78638 Merge pull request #14333 from asutoshpalai/arm-rbenv-fix
Error: Please update your system curl or set HOMEBREW_CURL_PATH to a newer version.
Minimum required version: 7.41.0
Your curl version:
Your curl executable:
Failed during: /home/linuxbrew/.linuxbrew/bin/brew update --force --quiet
```

![capture d'écran](https://user-images.githubusercontent.com/107509668/210813641-eed69238-7a7b-42b5-97f9-f3bf3c7ce355.jpg)

So at the beginning, it's well written that Git wad found, also CURL. But then, it seems that my curl version is not found finally. "Your curl version:" it's blanked. The same for "your curl executable". 

Any idea ? Thank in advance for your time.

**Reply by gromgit** [reply]
>```
> export HOMEBREW_CURL_PATH=/bin/curl
> export HOMEBREW_GIT_PATH=/bin/git
>```

Is there a reason you're using these values, instead of the ones I gave you earlier?

**Reply by leehanchung** [reply]
Running exactly into the same issue. Set the env vars to `whereis` and its still not finding it.

Ubuntu 20.04 running on WSL.

```bash
$ export HOMEBREW_GIT_PATH=/usr/bin/git
$ export HOMEBREW_CURL_PATH=/usr/bin/curl
$ env | grep HOMEBREW
HOMEBREW_GIT_PATH=/usr/bin/git
HOMEBREW_CURL_PATH=/usr/bin/curl
$ /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
==> Found Git: ./bin/git
==> Found cURL: ./bin/curl
==> Setting HOMEBREW_DEVELOPER to use Git/cURL not in /usr/bin
==> Checking for `sudo` access (which may request your password)...
==> This script will install:
/home/linuxbrew/.linuxbrew/bin/brew
/home/linuxbrew/.linuxbrew/share/doc/homebrew
/home/linuxbrew/.linuxbrew/share/man/man1/brew.1
/home/linuxbrew/.linuxbrew/share/zsh/site-functions/_brew
/home/linuxbrew/.linuxbrew/etc/bash_completion.d/brew
/home/linuxbrew/.linuxbrew/Homebrew

Press RETURN/ENTER to continue or any other key to abort:
==> /usr/bin/sudo /bin/chown -R han:han /home/linuxbrew/.linuxbrew/Homebrew
==> Downloading and installing Homebrew...
HEAD is now at 6db7732fa Merge pull request #14821 from issyl0/support-date-ranges-in-pr-reviews-query
Error: Please update your system curl or set HOMEBREW_CURL_PATH to a newer version.
Minimum required version: 7.41.0
Your curl version:
Your curl executable:
Failed during: /home/linuxbrew/.linuxbrew/bin/brew update --force --quiet
$ env | grep HOMEBREW
HOMEBREW_GIT_PATH=/usr/bin/git
HOMEBREW_CURL_PATH=/usr/bin/curl
```

**Reply by gromgit** [reply]
Note that the `HOMEBREW_{GIT,CURL}_PATH` variables are used by the `brew` command, not by the installer. The "Please update your system curl" error you see is actually coming from `brew`, and the installer _resets_ `HOMEBREW_{GIT,CURL}_PATH` internally to wherever it finds a suitable git/curl in your `$PATH`. It apparently found both in `./bin`, which is an _unstable path_.

In short:
1. Your system curl (and possibly git as well) is too old.
2. `./bin` in your `$PATH` is a **BIG NO-NO** for both practical (one `cd` makes your new git/curl inaccessible--and the installer does *four* of them) and security reasons (one haX0r can make you run a malware `ls` with zero effort).
3. There's a installer bug that fails to use the binaries that were actually found. PR in progress: https://github.com/Homebrew/install/pull/751

**Reply by leehanchung** [reply]
> 1. Your system curl (and possibly git as well) is too old.

It's asking for `curl > 7.41` and I have 7.68.  Latest git is 2.39 and I have 2.25.

```bash
:~$ curl -V
curl 7.68.0 (x86_64-pc-linux-gnu) libcurl/7.68.0 OpenSSL/1.1.1f zlib/1.2.11 brotli/1.0.7 libidn2/2.2.0 libpsl/0.21.0 (+libidn2/2.2.0) libssh/0.9.3/openssl/zlib nghttp2/1.40.0 librtmp/2.3
Release-Date: 2020-01-08
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp
Features: AsynchDNS brotli GSS-API HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL TLS-SRP UnixSockets
$ git --version
git version 2.25.1
```

> ./bin in your $PATH is a BIG NO-NO for both practical (one cd makes your new git/curl inaccessible--and the installer does four of them) and security reasons (one haX0r can make you run a malware ls with zero effort).

Thank you. I don't see that in my `~/.bashrc`. But will investigate and fix.

**Comment by osalbahr** [comment]
Were you able to fix the issue? I am having the same thing in #4544.

**Comment by edith-403** [comment]
I fixed this by simply installing curl:
sudo apt-get update
sudo apt-get install curl

And then installing Homebrew with:
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)

This solved the issue for me instead of using:
export HOMEBREW_CURL_PATH=$(which curl)

**Reply by osalbahr** [reply]
I'm curious, how were you able to get the error above without having `curl` installed from the first place? Perhaps `apt-get install curl` simply upgraded `curl`, i.e. you haven't ran `apt-get upgrade` for a while?

Because without `curl` installed, I get the following error instead of the `HOMEBREW_CURL_PATH`message:
```console
osalbahr@bb3b313f04c5:~/install$ ./install.sh 
...
You must install cURL before installing Homebrew. See:
  https://docs.brew.sh/Installation
```
(Tested in a `docker --rm -it ubuntu` with only `git` and `sudo` installed, and FWIW [Homebrew-on-Linux#requirements](https://docs.brew.sh/Homebrew-on-Linux#requirements) lists a few more)
